### PR TITLE
ENH: Add pandas_compat.table_from_frame(df)

### PR DESCRIPTION
**Pull request by kernc**
##### Issue
Various third parties work with pandas.DataFrames. Missing was a convenient method of transforming result DataFrame into Orange Table.

##### Description of changes
Add `table_from_frame` utility function. No hard pandas dependencies. Please advise on the best way to run the tests.

##### Includes
- [X] Code changes
- [x] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by astaric** [comment]
AFAIAC pandas can be installed on travis (and appveyor) to run the tests while not being added to the install_requires listed in setup.py.

I assume this is how pandas tests HDFS support as there are tests in their repository and tables listed in optional-requirements, while it does not get installed when you run `pip install pandas`.

File: Orange/data/tests/test_pandas.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+try:
+    import pandas
+except ImportError:
+    pandas = None
+
+@unittest.skipIf(pandas is None, "Missing package 'pandas'")
+class TestPandasCompat(unittest.TestCase):
+    def test_table_from_frame(self):
+        from Orange.data.pandas_compat import Test
+        Test().test_table_from_frame()
```
Comment (astaric) [inline Orange/data/tests/test_pandas.py]: Why don't you move the body of the test here?

File: Orange/data/pandas_compat.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+    Returns
+    -------
+    Table
+    """
+
+    def _is_discrete(s):
+        return (is_categorical_dtype(s) or
+                is_object_dtype(s) and (force_nominal or
+                                        s.nunique() < s.size**.666))
```
Comment (astaric) [inline Orange/data/pandas_compat.py]: I guess you do not want to use [is_discrete_values](https://github.com/biolab/orange3/blame/master/Orange/data/variable.py#L33) here as it is too complicated, but you could at least use the same constant :)

File: Orange/data/pandas_compat.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            s = pd.to_datetime(s, infer_datetime_format=True)
+            X.append(s.astype('str').replace('NaT', np.nan).map(tvar.parse).values)
+        elif is_numeric_dtype(s):
+            attrs.append(ContinuousVariable(name))
+            X.append(s.values)
+        else:
+            metas.append(StringVariable(name))
+            M.append(s.values.astype(object))
+
+    MAX_LENGTH = max(len(X[0]) if X else 0,
```
Comment (astaric) [inline Orange/data/pandas_compat.py]: Is this ever going to be different than df.shape[0] (or equivalent)?

File: Orange/data/pandas_compat.py (hunk 2 above)
Comment (kernc) [reply Orange/data/pandas_compat.py]: 👍

It's the exact same [universal] constant, just computed with more precision. 😛

File: Orange/data/pandas_compat.py (hunk 3 above)
Comment (kernc) [reply Orange/data/pandas_compat.py]: Constructing Table objects, one sometimes gets so invested with details that they tend to lose grasp of the whole picture.

Thanks.

**Comment by astaric** [comment]
The code look ok to me. Squash if you intended to and I'll merge.

Since you ticked the documentation checkbox, should this be mentioned in the docs somewhere? I do not have a specific place in mind so the aswer to this could be "no", or "not yet", but I think it is a useful helper more people should know about :)

**Comment by kernc** [comment]
Was referring to reasonable docstring as provided. Guessing reviewers should be the ones to catch it until DataFrames become a more frequently integrated pattern.

**Pull request merged**